service/service.go

```diff
@@ -71,24 +79,19 @@ type Service struct {
 
 	// openRound is the round which is currently open for accepting challenges registration from miners.
 	// At any given time there is one single open round.
-	// openRoundMutex guards openRound, any access to it must be protected by this mutex.
-	openRound      *round
-	openRoundMutex sync.RWMutex
-
-	// executingRounds are the rounds which are currently executing, hence generating a proof.
-	executingRounds      map[string]*round
-	executingRoundsMutex sync.RWMutex
+	openRound       *round
+	executingRounds map[string]struct{}
 
 	PubKey  ed25519.PublicKey
 	privKey ed25519.PrivateKey
 
-	broadcaster       atomic.Value // holds Broadcaster interface
-	challengeVerifier atomic.Value // holds challenge_verifier.Verifier
+	broadcaster Broadcaster
 
-	errChan chan error
 	sync.Mutex
 }
 
+type Command func(*Service)
```

Comment:
Please add some documentation to `Command`, e.g. that the command must terminate in a timely manner and should not do file I/O.

Imo a command should also receive the context of the function that actually executes it, so it can return early if the work it does isn't required anymore.

Reply:
Do you mean the context of the function that creates and schedules the `Command` or the context of `Service::loop`? The former is already available as a variable captured by a closure.  If I pass the context of `Service::loop` there will be two contexts available which will be confusing.

Reply:
The context of the caller of Command, not the creator of it.

```golang
type Command func(ctx context.Context, *Service)

// creator
cmdChan <- func(ctx context.Context, s *Service) {
	select {
	case <- ctx.Done():
		return
	default:
	}

	s.variable = newValue
}

// caller (b.loop)
cmd <- cmdChan

cmd(ctx, s) // ctx here is the context that cancels loop
```

Reply:
It will be confusing because it will look like this:
```go
func (s *ServiceClient) Info(ctx context.Context) (*InfoResponse, error) {
	s.command <- func(serviceCtx context.Context s *Service) {
		// THERE ARE 2 Contexts here, it's confusing
		select {
		case <-ctx.Done():
			return // Info's context is cancelled - GRPC is cancelled
		case <- serviceCtx.Done():
			return // Service's context is cancelled - shutting down
		default:
		}
		s.variable = newValue
	}
}
```

Relying on the requirement that `Command` is supposed to be **short**, I don't see a need to select on any of these. It should just assign `newValue`  / run and return.

Reply:
The code you posted is wrong. A `Command` should never check the context of the function that created it. That doesn't make any logical sense.

With the same argument I could say that a command shouldn't have a `Service` parameter, because it already has access to the service from it's parents scope.

That being sad: I think `ServiceClient` isn't required as a type. imo inlining it into `Service` makes the code much more straight forward than having an embedded service client.

Reply:
Like this: https://github.com/spacemeshos/poet/compare/170-refactor-service-to-actor...inline-service-client

Code is still thread save but all the nesting is removed.

Reply:
The point of extracting a `ServiceClient` is to prevent accidental access to the state of `Service` without synchronization. It is just not possible and guaranteed at compile time. It is to make the code more robust and to prevent mistakes leading to data races. This guard is lost when the `Info` and `Submit` methods are integrated back into the `Service` as they can now access the state of `Service` without any synchronization.

Reply:
But what ensures that any method that changes the `Service` isn't part of `Service` but part of `ServiceClient`?

Reply:
The key point of the actor pattern is that all access to the state of the actor (`Service`) is done in a single goroutine.

After this refactoring, there isn't any goroutine, other than the one running `Serivce::loop`, that has a reference to the `Service`. Thanks to this all mutations and reads of the state are synchronized. 

It's perfectly fine if there is another private method of `Service` that accesses its state as there is only one goroutine that can call them (the one running the `Service::loop()`. The `Service` must expose all its public API thru the `ServiceClient` to make the assumptions hold.

Reply:
reading service.go after your change, the differentiation between `Service` and `ServiceClient` is still unclear to me.
i don't understand why ServiceClient cannot be inlined.
it seems the public APIs Info/Submit/SetBroadcaster/SetChallengeVerifier all can submit through the command channel.

Reply:
`ServiceClient` _could_ be inlined into `Service`. Not inlining it prevents a whole class of bugs as there is literally no way to access the `Service` state from `ServiceClient` in any other way than safely via the `commands` channel. Thanks to this, it is impossible to accidentally create a data race from the `Submit` and `Info` methods.

Reply:
need to start the loop when service is created too

Reply:
Thanks, @countvonzero. I'm going to move the start of the event loop to where the `Service` is created, which will allow me to remove the last mutex. I'm going to do this in a PR removing broadcasting as this is the only reason for delaying the Service start (it cannot operate w/o a broadcaster).